File: examples/order/ts/src/providers.ts
import type { Order, OrderID } from "./domain.js";

// A Provider is an Interface the Core Declares
// and something outside Fulfils. See RULES.md, Providers.

// Page Carries Pagination without Naming a Database.
export type Page = { number: number; size: number };

// OrderStore Keeps Orders wherever Orders Live.
// Fulfilled by SqliteOrders.
// Reference: https://nodejs.org/api/sqlite.html
export interface OrderStore {
  insertOrderRow(order: Order): Order;
  selectOrderRow(id: OrderID): Order;
  selectOrderPage(page: Page): Order[];
}

File: examples/order/go/providers.go
package main

// A Provider is an Interface the Core Declares
// and something outside Fulfils. See RULES.md, Providers.

// OrderStore Keeps Orders wherever Orders Live.
// Fulfilled by GormOrders.
// Reference: https://gorm.io/docs/
type OrderStore interface {
	InsertOrderRow(o Order) (Order, error)
	SelectOrderRow(id OrderID) (Order, error)
	SelectOrderPage(page Page) ([]Order, error)
}

// Page Carries Pagination without Naming a Database.
type Page struct {
	Number int
	Size   int
}

// CheckPageBounds Refuses a Page the Store cannot Serve.
func (p Page) CheckPageBounds() error {
	if p.Number < 0 || p.Size < 0 {
		return ErrPageIsInvalid
	}

	return nil
}
